### ISSUE TITLE
Improve startup time

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/fd#1408]
`fd`s startup time is quite slow. On my 12 core system, it takes ~ 20 ms for "searching" an empty folder. This is fast enough not to be noticeable by humans, but it looks bad in benchmarks when comparing `fd` with other tools on small folders [^1]. And it's also an actual problem for use cases where `fd` is called repeatedly from a script.

Some of that overhead is caused by the spawning of threads, and that problem is already tracked in https://github.com/sharkdp/fd/issues/1203. But I think there is more that can be done. Instead of using my usual go-to performance tool (`perf`), let's look at the [magic-trace](https://github.com/janestreet/magic-trace) output of a `fd` call in an empty folder [^2]. If someone is interested, I've attached the [trace](https://github.com/sharkdp/fd/files/13079377/trace.zip) to this post. Go to https://magic-trace.org/ to load it in their viewer.

The full trace looks like this:

![image](https://github.com/sharkdp/fd/assets/4209276/14619eb4-2e1a-4f5a-9f3a-122aa6438173)

The first 2.2 ms are typical process startup things (before `main`). I don't think there is any room for optimization here (?)

![image](https://github.com/sharkdp/fd/assets/4209276/c13a7b28-d22f-40a1-a8e5-104dd6fb16f5)

The next ~2 ms are more interesting:

![image](https://github.com/sharkdp/fd/assets/4209276/51ede8bd-883e-4031-ad6d-0f2934b66c14)

some notable steps (even if insignificant in time)
- parsing command-line arguments (**724 µs**)
- a "is_existing_directory" check on the search path (28 µs)
- the parsing of the (empty) search pattern regex (32 µs)
- `isatty` check (5 ·s)
- `LsColors::from_env` (**579 µs**)
- num_cpus::linux::get_num_cpus` (**441 µs**)
- RegexBuilder::build (171 µs)

Some things were surprising to me. I didn't expect the `get_num_cpus` call to take this long. There might be some room for improvement here by doing things in parallel (e.g. `LsColors::from_env`)? But only if the thread overhead is not too high.

Then we start the actual scan, which takes the majority of the time:

![image](https://github.com/sharkdp/fd/assets/4209276/5dc31bf4-7771-448e-9bee-b6895ced1c5e)

Here, I'm not so sure how to interpret the trace, as things are actually happening on multiple threads. But we can (presumably) see some of the thread spawning/joining time here (~ **5 ms**):

![image](https://github.com/sharkdp/fd/assets/4209276/578a922e-3656-4966-90b9-c37eec18b64d)

and some gitignore matcher logic going on here (370 µs total):

![image](https://github.com/sharkdp/fd/assets/4209276/400de973-fb8f-4e63-9411-4140e9a4d7c1)

Most of the time is actually unaccounted for in the trace, because I can only see:

![image](https://github.com/sharkdp/fd/assets/4209276/253f364a-0a8d-46d8-9fb9-e82af7ece87d)

We can see a bit more when switching off LTO:

![image](https://github.com/sharkdp/fd/assets/4209276/6741b564-99e3-4565-b2d3-1d49d062663b)

Apparently, **11 ms** are spent in `crossbeam_channel::channel::bounded`'s `from_iter` method? (probably the receive call?) — even though we don't have any work to do. On a `-j1` run, this part only takes 1 ms.

[^1]: Those "small" folders can be pretty large, actually. It takes hundreds of thousands of files before we can make up for the startup "penalty".
[^2]: I recently discovered this and used it successfully to benchmark (and then optimize) the *startup* time of other programs.

**Comment by tmccombs** [comment]
> 11 ms are spent in crossbeam_channel::channel::bounded's from_iter method?

That might be where crossbeam_channel initializes the memory for the channel
[here](https://github.com/crossbeam-rs/crossbeam/blob/9dd4c9594dd7c84abf04e89192f1bdfe790100f5/crossbeam-channel/src/flavors/array.rs#L116).
